external-crates/move/crates/move-compiler/src/linters/self_assignment.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Detects and reports explicit self-assignments in code, such as `x = x;`, which are generally unnecessary
+//! and could indicate potential errors or misunderstandings in the code logic.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::{
+        ast::{self as T, LValue_, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+use super::{LinterDiagCategory, LINTER_SELF_ASSIGNMENT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+const SELF_ASSIGNMENT_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::Suspicious as u8,
+    LINTER_SELF_ASSIGNMENT_DIAG_CODE,
+    "Explicit self-assignment detected. This operation is usually unnecessary and could indicate a typo or logical error.",
+);
+pub struct SelfAssignmentCheck;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for SelfAssignmentCheck {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::Assign(value_list, _, assign_exp) = &exp.exp.value {
+            if let Some(
+                sp!(
+                    _,
+                    LValue_::Var {
+                        var: sp!(_, lhs),
+                        ..
+                    }
+                ),
+            ) = value_list.value.get(0)
+            {
+                if let UnannotatedExp_::Copy {
+                    var: sp!(_, rhs), ..
+                } = &assign_exp.exp.value
```

Comment:
Should this also handle the `Move` case?

Reply:
refactored